lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,353 @@
+	Version               uint32 /* audit api version number or feature bitmap */
+	BacklogWaitTime       uint32 /* message queue wait timeout */
+	BacklogWaitTimeActual uint32 /* message queue wait timeout */
+}
+
+// IsLoginUIDSet returns true if login UID is set, false otherwise.
+func IsLoginUIDSet() bool {
+	if !hasCapabilities() {
+		// Current process doesn't have system permissions to talk to auditd.
+		return false
+	}
+
+	client := NewClient(Message{})
+	defer func() {
+		if err := client.Close(); err != nil {
+			log.WithError(err).Warn("failed to create auditd client")
+		}
+	}()
+	// We don't need to acquire the internal client mutex as the connection is
+	// not shared.
+	if err := client.connectUnderMutex(); err != nil {
+		return false
+	}
+
+	enabled, err := client.isEnabledUnderMutex()
+	if err != nil || !enabled {
+		return false
+	}
+
+	loginuid, err := getSelfLoginUID()
+	if err != nil {
+		log.WithError(err).Debug("failed to read login UID")
+		return false
+	}
+
+	// if value is not set, logind PAM module will set it to the correct value
+	// after fork. 4294967295 (math.MaxUint32) is -1 converted to uint32.
+	return loginuid != math.MaxUint32
+}
+
+func getSelfLoginUID() (int64, error) {
+	data, err := os.ReadFile("/proc/self/loginuid")
+	if err != nil {
+		return 0, trace.ConvertSystemError(err)
+	}
+
+	loginuid, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return loginuid, nil
+}
+
+// SendEvent sends a single auditd event. Each request create a new netlink connection.
+// This function does not send the event and returns no error if it runs with no root permissions.
+func SendEvent(event EventType, result ResultType, msg Message) error {
+	if !hasCapabilities() {
+		// Do nothing when not running as root.
+		return nil
+	}
+
+	client := NewClient(msg)
+	defer func() {
+		err := client.Close()
+		if err != nil {
+			log.WithError(err).Error("failed to close auditd client")
+		}
+	}()
+
+	if err := client.SendMsg(event, result); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (c *Client) connectUnderMutex() error {
+	if c.conn != nil {
+		// Already connected, return
+		return nil
+	}
+
+	conn, err := c.dial(syscall.NETLINK_AUDIT, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	c.conn = conn
+
+	return nil
+}
+
+func (c *Client) isEnabledUnderMutex() (bool, error) {
+	if c.enabled != unset {
+		// We've already gotten the status.
+		return c.enabled == enabled, nil
+	}
+
+	status, err := getAuditStatus(c.conn)
+	if err != nil {
+		return false, trace.Errorf("failed to get auditd status: %v", trace.ConvertSystemError(err))
+	}
+
+	// enabled can be either 1 or 2 if enabled, 0 otherwise
+	if status.Enabled > 0 {
+		c.enabled = enabled
+	} else {
+		c.enabled = disabled
+	}
+
+	return c.enabled == enabled, nil
+}
+
+// NewClient creates a new auditd client. Client is not connected when it is returned.
+func NewClient(msg Message) *Client {
```

Comment:
It is interesting that the client is created per message. Is there a reason it can't accepts message in the "send" method?

Reply:
Yes, here is the discussion about that https://github.com/gravitational/teleport/pull/14948/files#r953974754